### PR TITLE
[GOBBLIN-691] Make format-specific component pluggable in compaction 

### PR DESCRIPTION
**Pull request by autumnust**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/691) issues and references them in the PR title. 
    - https://issues.apache.org/jira/browse/GOBBLIN-691

### Description
- As the first step of ORC compaction, this PR tries to make current compaction code insensitive to the data format: Format-specific components should be pluggable: `CompactionJobConfigurator ` becomes the abstract implementation for `CompactionAvroJobConfigurator ` which used to be the component for Avro-specific logic, so that we can implement other formats' configurator to use the same dataset discovery and verification logic.  All changes should be backward compatible. 
- Put `deprecation` annotation for old compaction implementation. Instead, we should use `CompactionSource` + `MRCompactionTask` to launch MR job. 
- Fix some of typo while reading thru. code. 

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by autumnust** [comment]
@yukuai518 @ibuenros Can you help review ? Thanks.

File: gobblin-compaction/src/main/java/org/apache/gobblin/compaction/CompactorFactory.java [hunk 1]
```diff
@@ -29,6 +29,7 @@
 
 /**
  * A factory responsible for creating {@link Compactor}s.
+ * @deprecated
```
Comment (ibuenros) [inline gobblin-compaction/src/main/java/org/apache/gobblin/compaction/CompactorFactory.java]: The annotation on javadoc is intended to explain reasons / alternatives to this API (which you should do). The actual deprecation annotation goes outside of the javadoc (`@Deprecated`)

File: gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/CompactionAvroJobConfigurator.java [hunk 2]
```diff
@@ -138,25 +106,27 @@ private Schema getKeySchema(Job job, Schema topicSchema) throws IOException {
...
-              + " not provided. Using key attributes in the schema for compaction");
+          + " not provided. Using key attributes in the schema for compaction");
       keySchema = AvroUtils.removeUncomparableFields(MRCompactorAvroKeyDedupJobRunner.getKeySchema(topicSchema)).get();
     }
 
     return keySchema;
   }
 
-  private void configureSchema(Job job) throws IOException {
+  protected void configureSchema(Job job) throws IOException {
```
Comment (ibuenros) [inline gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/CompactionAvroJobConfigurator.java]: Please add `@Override` annotation.

File: gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/CompactionJobConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+      job.setNumReduceTasks(0);
+    }
+    // Configure schema at the last step because FilesInputFormat will be used internally
+    this.configureSchema(job);
+    this.isJobCreated = true;
+    this.configuredJob = job;
+    return job;
+  }
+
+  protected abstract void configureSchema(Job job) throws IOException;
```
Comment (ibuenros) [inline gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/CompactionJobConfigurator.java]: Please add javadoc for all overridable methods.

**Comment by yukuai518** [comment]
+1 LGTM.

**Pull request closed without merging**